3.13/specializer.md

```diff
@@ -0,0 +1,289 @@
+evaluates the bodies of pure instructions with their constant values.
+We get this for free.
+
+### Guard elimination
+We eliminate guards when we see they are redundant using constant
+or type information.
+
+### True function inlining
+
+CPython 3.11 obtained frame inlining. This eliminated most of the
+overhead of function calls. However, "fake frames"
+were still needed. These are the instructions `_PUSH_FRAME` and `_POP_FRAME`.
+A similar idea was introduced in Cinder (Instagram's fork of CPython).
+
+In CPython 3.13, we plan for true function inlining -- that is, no
+frame creation at all in the happy case, not even fake frames.
+
+The uops tracer already projects traces through Python function boundaries.
+The goal is to eliminate `_PUSH_FRAME` and `_POP_FRAME`. Doing this is easy,
+but doing this while maintaining CPython semantics is very hard.
+
+The goal is to change this trace of instructions:
+```
+ADD_TO_TRACE(_CHECK_PEP_523, 1, 0)
+ADD_TO_TRACE(_CHECK_FUNCTION_EXACT_ARGS, 1, 4828)
+ADD_TO_TRACE(_CHECK_STACK_SPACE, 1, 0)
+# Frame creation
+ADD_TO_TRACE(_INIT_CALL_PY_EXACT_ARGS, 1, 0)
+ADD_TO_TRACE(_SAVE_RETURN_OFFSET, 4, 0)
+ADD_TO_TRACE(_PUSH_FRAME, 1, 0)
+ADD_TO_TRACE(_SET_IP, 0, 0)
+ADD_TO_TRACE(_CHECK_VALIDITY, 0, 0)
+ADD_TO_TRACE(_RESUME_CHECK, 0, 0)
+...
+# Frame removal
+ADD_TO_TRACE(_POP_FRAME, 0, 0)
+```
+to this:
+```
+_CHECK_PEP_523(1, 12, 0)
+_CHECK_FUNCTION_EXACT_ARGS(1, 12, 4828)
+_CHECK_STACK_SPACE(1, 12, 0)
+_SETUP_TIER2_FRAME(128, 0, 0)
+ # Minimal bookkeeping, no frame creation
+_PRE_INLINE(7, 0, 39)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+...
+_POST_INLINE(4, 0, 18446744073709551615)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+```
+
+
+#### Locals interleaving
+
+By interleaving the locals of the new call with the stack of the old
+function frame, we can achieve zero argument copying.
+
+Before a function call, the stack looks like this:
+
+| callable | self_or_null | arg1 | arg2              |
+|----------|--------------|------|-------------------|
+|          |              |      | ^ (stack pointer) |
+
+Notice that the arguments are already layed out in how the locals
+of the new "frame" will be. Thus, we just need to expand the stack
+to accomodate for more locals. Finally, since we are now pointing
+into the current stack as our locals, we offset all `LOAD_FAST`
+and `STORE_FAST` into the current stack.
+
+| callable | self_or_null                              | arg1 | arg2 |                   |
+|----------|-------------------------------------------|------|------|-------------------|
+|          | ^ `LOAD_FAST 0 + offset` of inlined frame |      |      | ^ (stack pointer) |
+
+Finally, we must update the attribute names' tuple that the current 
+frame points to, so that the inlined call's attribute loads load
+the correct attribute names. This is just a simple assignment.
+
+#### Compatiblity
+We aim for 100% compatibility with all existing CPython code, and
```

Comment:
This should go unsaid. 100% compatibility is a requirement for any T2 optimizer.